carma_cooperative_perception/src/multiple_object_tracker_component.cpp

```diff
@@ -0,0 +1,446 @@
+  carma_cooperative_perception_interfaces::msg::Track msg;
+
+  msg.header.stamp.sec = mot::remove_units(units::math::floor(track.timestamp));
+  msg.header.stamp.nanosec = mot::remove_units(
+    units::time::nanosecond_t{units::math::fmod(track.timestamp, units::time::second_t{10.0})});
+  msg.header.frame_id = "map";
+
+  msg.id = track.uuid.value();
+  msg.motion_model = msg.MOTION_MODEL_CTRA;
+  msg.pose.pose.position.x = mot::remove_units(track.state.position_x);
+  msg.pose.pose.position.y = mot::remove_units(track.state.position_y);
+
+  tf2::Quaternion orientation;
+  orientation.setRPY(0, 0, mot::remove_units(track.state.yaw.get_angle()));
+  msg.pose.pose.orientation.x = orientation.getX();
+  msg.pose.pose.orientation.y = orientation.getY();
+  msg.pose.pose.orientation.z = orientation.getZ();
+  msg.pose.pose.orientation.w = orientation.getW();
+
+  msg.twist.twist.linear.x = mot::remove_units(track.state.velocity);
+  msg.twist.twist.angular.z = mot::remove_units(track.state.yaw_rate);
+
+  msg.accel.accel.linear.x = mot::remove_units(track.state.acceleration);
+
+  return msg;
+}
+
+static auto to_ros_msg(const mot::CtrvTrack & track) noexcept
+{
+  carma_cooperative_perception_interfaces::msg::Track msg;
+
+  msg.header.stamp.sec = mot::remove_units(units::math::floor(track.timestamp));
+  msg.header.stamp.nanosec = mot::remove_units(
+    units::time::nanosecond_t{units::math::fmod(track.timestamp, units::time::second_t{10.0})});
+  msg.header.frame_id = "map";
+
+  msg.id = track.uuid.value();
+  msg.motion_model = msg.MOTION_MODEL_CTRV;
+  msg.pose.pose.position.x = mot::remove_units(track.state.position_x);
+  msg.pose.pose.position.y = mot::remove_units(track.state.position_y);
+
+  tf2::Quaternion orientation;
+  orientation.setRPY(0, 0, mot::remove_units(track.state.yaw.get_angle()));
+  msg.pose.pose.orientation.x = orientation.getX();
+  msg.pose.pose.orientation.y = orientation.getY();
+  msg.pose.pose.orientation.z = orientation.getZ();
+  msg.pose.pose.orientation.w = orientation.getW();
+
+  msg.twist.twist.linear.x = mot::remove_units(track.state.velocity);
+  msg.twist.twist.angular.z = mot::remove_units(track.state.yaw_rate);
+
+  return msg;
+}
+
+static auto to_ros_msg(const Track & track)
+{
+  static constexpr mot::Visitor visitor{
+    [](const mot::CtrvTrack & t) { return to_ros_msg(t); },
+    [](const mot::CtraTrack & t) { return to_ros_msg(t); },
+    [](const auto &) { throw std::runtime_error{"cannot make ROS 2 message from track type"}; }};
+
+  return std::visit(visitor, track);
+}
+
+MultipleObjectTrackerNode::MultipleObjectTrackerNode(const rclcpp::NodeOptions & options)
+: CarmaLifecycleNode{options}, track_manager_{mot::PromotionThreshold{3}, mot::RemovalThreshold{0}}
+{
+  // CarmaLifecycleNode base class will automatically handle lifecycle state changes for
+  // lifecycle publishers and timers.
+  lifecycle_publishers_.push_back(track_list_pub_);
+  timers_.push_back(pipeline_execution_timer_);
```

Comment:
I'm not sure how else the parent class would know about timers and publishers. I could be missing something, but if other derivative classes are not appending to the `lifecycle_publishers_` and `timers_` `std::vector`s, then they are doing it wrong.

Reply:
it does look like the overrides of the functions does it
https://github.com/usdot-fhwa-stol/carma-utils/blob/a2cb00954c8ba2e4837e1b16e6cad55457914bdd/carma_ros2_utils/src/carma_lifecycle_node.cpp#L229

Reply:
The parent class updates the lifecycle states for timers and lifecycle publishers when the containing node changes state, but the `lifecycle_publishers_` `std::vector` is initially empty. We need to populate the `std::vector`, otherwise, the `for` loops in the `*_publishers()` functions won't do anything.

Reply:
After looking back through the `CarmaLifecycleNode` class implementation, I see where the timers and lifecycle publishers get added to their respective lists.